src/ai/backend/manager/server.py

```diff
@@ -495,6 +496,119 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
         await ectx.cleanup()
 
 
+@actxmgr
+async def hanging_session_scanner_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    from contextlib import suppress
+    from datetime import timedelta
+    from typing import TYPE_CHECKING
+
+    import sqlalchemy as sa
+    from dateutil.relativedelta import relativedelta
+    from dateutil.tz import tzutc
+    from sqlalchemy.orm import load_only, noload
+
+    from .config import session_hang_tolerance_threshold_iv
+    from .models.session import SessionStatus
+
+    if TYPE_CHECKING:
+        from .models.utils import ExtendedAsyncSAEngine
+
+    async def _fetch_hanging_sessions(
+        db: ExtendedAsyncSAEngine,
+        status: SessionStatus,
+        threshold: timedelta,
+    ) -> tuple[SessionRow, ...]:
+        query = (
+            sa.select(SessionRow)
+            .where(SessionRow.status == status)
+            .where(
+                (
+                    datetime.now(tz=tzutc())
+                    - SessionRow.status_history[status.name].astext.cast(
+                        sa.types.DateTime(timezone=True)
+                    )
+                )
+                > threshold
+            )
+            .options(
+                noload("*"),
+                load_only(SessionRow.id, SessionRow.name, SessionRow.status, SessionRow.access_key),
+            )
+        )
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            return result.fetchall()
+
+    async def _force_terminate_hanging_sessions(
+        status: SessionStatus,
+        threshold: timedelta,
+    ) -> None:
+        heuristic_weight = 0.4  # NOTE: Shorter than a half(0.5)
+        heuristic_interval = threshold.total_seconds() * heuristic_weight
+        while True:
+            try:
+                sessions = await _fetch_hanging_sessions(
+                    root_ctx.db, status=status, threshold=threshold
+                )
+            except Exception as e:
+                log.error(e)
+                await asyncio.sleep(10)
+                continue
+
+            log.debug(f"{len(sessions)} {status.name} sessions found.")
+
+            results_and_exceptions = await asyncio.gather(
+                *[
+                    asyncio.create_task(
+                        root_ctx.registry.destroy_session(
+                            session, forced=True, reason=KernelLifecycleEventReason.HANG_TIMEOUT
+                        ),
+                    )
+                    for session in sessions
+                ],
+                return_exceptions=True,
+            )
+            for result_or_exception in results_and_exceptions:
+                if isinstance(result_or_exception, (BaseException, Exception)):
+                    log.error(result_or_exception)
+
+            await asyncio.sleep(heuristic_interval)
```

Comment:
It would be better to use aiotools timer to split out what it does on each interval and how we invoke it upon each interval. (e.g., separation of the unhandled exception from the timer loop)